manifests/server/config/user.pp

```diff
@@ -0,0 +1,120 @@
+# @summary Define ClickHouse users
+#
+# This type defines the ClickHouse users with specified parameters
+#
+# @param networks
+#   Defines network nodes where from user could connect
+#
+# @param profile
+#   User profile
+#
+# @param quota
+#   User quota
+#
+# @param user
+#   User name
+#
+# @param password
+#   Plaintext password. Will be hashed into sha256 format in xml file
+#
+# @param password_sha256
+#   Optional parameter. If defined, it will be used despite the value of `password`
+#
+# @param databases
+#   Optional hash of allowed DBs, tables and filters to implement ACLs
+#
+# @param allow_databases
+#   Optional parameter to restrict access to specified databases
+#
+# @param allow_dictionaries
+#   Optional parameter to restrict access to specified dictionaries
+#
+# @see https://clickhouse.tech/docs/en/operations/settings/settings_users/, https://clickhouse.tech/docs/v20.3/en/operations/access_rights/
+#
+# @example
+#   clickhouse::server::config::user { 'username':
+#       network   => {
+#           'ip'          => [
+#               '::',
+#               '0.0.0.0',
+#           ],
+#           'host'        => [
+#               'host1.local',
+#               'host2.local',
+#           ],
+#           'host_regexp' => [
+#               '[^.]*\.domain\.TLD',
+#           ],
+#       },
+#       profile   => 'profile_name',
+#       quota     => 'quota_name',
+#       password  => 'password',
+#       databases => {
+#           'db_name' => {
+#               'table_name'         => ['filter'],
+#               'another_table_name' => ['another filter'],
+#           },
+#       },
+#   }
+#
+# @author InnoGames GmbH
+#
+define clickhouse::server::config::user (
+    Struct[{
+        Optional[ip]          => Array[String[1], 1],
+        Optional[host]        => Array[String[1], 1],
+        Optional[host_regexp] => Array[String[1], 1],
+    }]                         $networks,
+    String[1]                  $profile            = 'default',
+    String[1]                  $quota              = 'default',
+    String[1]                  $user               = $title,
+    String[0]                  $password           = '',
```

Comment:
will look there closer, but the empty password is completely fine, so I wouldn't force it

Reply:
So. Here's a logic:

- `$password` could be empty, it's completely fine
- If `$password_sha256` is defined and correct (and it's checked by Pattern[]), then it will be used https://github.com/innogames/puppet-clickhouse/pull/4/files/5f9e622695c2657eba90013d0f5c036dc206685e..345e10918638e193fffc01a867d1ce84380f8617#diff-4d46b5ba64830bec4a95cf232cfcee42R85
- If not, `sha256($password)` is used.

I don't see the need of additional checks. Is it?

Reply:
a clean api does not allow user to do unsupported things. if the user sets both password parameters they can be different. the user must decide whether we wants to provide a plaintext or a hashed password. if you say an empty password is fine then just check that not both parameters are set

Reply:
Fair enough. Ok, it makes sense to check if not both are set.